File: Src/Compilers/CSharp/Source/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.ConstructedTypeSymbol.cs
//-----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections.Concurrent;
using System.Diagnostics;

namespace Roslyn.Compilers.CSharp
{
    internal sealed partial class AnonymousTypeManager
    {
        /// <summary>
        /// Represents a constructed instance of an anonymous type 'template'.
        /// </summary>
        private sealed class ConstructedAnonymousTypeSymbol : ConstructedNamedTypeSymbol
        {
            internal ConstructedAnonymousTypeSymbol(NamedTypeSymbol constructedFrom, ReadOnlyArray<TypeSymbol> typeArguments, AnonymousTypeDescriptor typeDescr)
                : base(constructedFrom, typeArguments)
            {
                this.TypeDescr = typeDescr;
            }

            internal readonly AnonymousTypeDescriptor TypeDescr;

            public override bool IsImplicitlyDeclared
            {
                get { return false; }
            }

            public override bool IsAnonymousType
            {
                get { return true; }
            }

            public override ReadOnlyArray<Location> Locations
            {
                get { return ReadOnlyArray<Location>.CreateFrom(this.TypeDescr.Location); }
            }

            public override ReadOnlyArray<SyntaxNode> DeclaringSyntaxNodes
            {
                get
                {
                    return GetDeclaringSyntaxNodeHelper<AnonymousObjectCreationExpressionSyntax>(this.Locations);
                }
            }

            private AnonymousTypeTemplateSymbol Template
            {
                get { return (AnonymousTypeTemplateSymbol)this.ConstructedFrom; }
            }

            public override string Name
            {
                get { return string.Empty; }
            }

            public override string MetadataName
            {
                get { return string.Empty; }
            }
            
            /// <summary> Adjust the smallest location in the 
            /// anonymous type template with this type's location </summary>
            internal void AdjustSmallestLocationInTemplate()
            {
                Debug.Assert(this.TypeDescr.Location.IsInSource);
                Template.AdjustLocation(this.TypeDescr.Location);
            }
        }
    }
}

File: Src/Compilers/VisualBasic/Source/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType_ConstructedTypeSymbol.vb
Imports System.Collections.Generic
Imports Roslyn.Compilers
Imports Roslyn.Compilers.Common
Imports Roslyn.Compilers.Internal

Namespace Roslyn.Compilers.VisualBasic

    Partial Friend NotInheritable Class AnonymousTypeManager

        Friend NotInheritable Class AnonymousTypeConstructedTypeSymbol
            Inherits AnonymousTypeOrDelegateConstructedTypeSymbol

            Public Sub New(substitution As TypeSubstitution, typeDescriptor As AnonymousTypeDescriptor)
                MyBase.New(substitution, typeDescriptor)
            End Sub

            Protected Overrides Function CompleteInternalSubstituteTypeParameters(newSubstitution As TypeSubstitution) As TypeSymbol
                Return New AnonymousTypeConstructedTypeSymbol(newSubstitution, TypeDescriptor)
            End Function

            Protected Overrides Function CreateSubstitutedPropertySymbol(memberProperty As PropertySymbol,
                                                                         getMethod As SubstitutedMethodSymbol,
                                                                         setMethod As SubstitutedMethodSymbol) As SubstitutedPropertySymbol
                Return New AnonymousTypeSpecializedPropertySymbol(Me, memberProperty, getMethod, setMethod,
                                                                  Me.TypeDescriptor.Fields(DirectCast(memberProperty, AnonymousTypePropertySymbol).PropertyIndex))
            End Function

            Protected Overrides Function SubstituteTypeParametersForMemberMethod(memberMethod As MethodSymbol) As SubstitutedMethodSymbol
                If memberMethod.MethodKind = MethodKind.PropertyGet OrElse memberMethod.MethodKind = MethodKind.PropertySet Then
                    Debug.Assert(memberMethod.Arity = 0)
                    Return New AnonymousTypeSpecializedPropertyAccessorSymbol(Me, memberMethod)

                ElseIf memberMethod.Parameters.Count = 1 AndAlso memberMethod.Parameters(0).Type Is Me.OriginalDefinition Then
                    Dim iEquatable_Equals = TryCast(memberMethod, AnonymousType_IEquatable_EqualsMethodSymbol)
                    Debug.Assert(iEquatable_Equals IsNot Nothing)

                    If iEquatable_Equals IsNot Nothing Then
                        Return New AnonymousTypeSpecialized_IEquatable_EqualsMethodSymbol(Me, iEquatable_Equals)
                    End If
                End If

                Return MyBase.SubstituteTypeParametersForMemberMethod(memberMethod)
            End Function

            Friend Overrides Function MakeAcyclicInterfaces(diagnostics As DiagnosticBag) As ReadOnlyArray(Of NamedTypeSymbol)
                Dim instanceInterfaces = OriginalNamedTypeDefinition.Interfaces

                If instanceInterfaces.Count = 0 Then
                    Return ReadOnlyArray(Of NamedTypeSymbol).Empty

                ElseIf instanceInterfaces.Count = 1 AndAlso instanceInterfaces(0).TypeArguments.Count = 1 AndAlso
                    instanceInterfaces(0).TypeArguments(0) Is Me.OriginalDefinition Then
                    ' Must be IEquitable(Of T).
                    Return ReadOnlyArray.Singleton(Of NamedTypeSymbol)(instanceInterfaces(0).OriginalNamedTypeDefinition.Construct(Me))
                End If

                Throw Contract.Unreachable
            End Function

            Public Overrides ReadOnly Property DeclaringSyntaxNodes As ReadOnlyArray(Of SyntaxNode)
                Get
                    Return GetDeclaringSyntaxNodeHelper(Of AnonymousObjectCreationExpressionSyntax)(ReadOnlyArray(Of Location).CreateFrom(Me.Locations))
                End Get
            End Property

            Public Overrides ReadOnly Property IsImplicitlyDeclared As Boolean
                Get
                    Return False
                End Get
            End Property
        End Class

    End Class

End Namespace